### PR TITLE
Reader

### PR DESCRIPTION
**Pull request by al7aro**

### CODE REVIEW

File: include/macro.h [hunk 1]
```diff
@@ -20,6 +20,10 @@
 # define NEW_LINE_CHAR		'\n'
 # define SINGLE_QUOTE_CHAR	'\''
 # define DOUBLE_QUOTE_CHAR	'\"'
+# define SPACE_CHAR			' '
+# define MAIN_PROMPT		"prompt> "
+# define DQUOTE_PROMPT		"dquote> "
```
Comment (arikhativa) [inline include/macro.h:25]: i think there should be a 'quote' promt
you dont need to do that in this PR, you can create a new PR for that

image

File: include/reader.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+# include "error_code.h"
+# include "libft.h"
+# include "tab.h"
+# include "macro.h"
+
+t_error_code	reader_split_arg(char *str, char ***ret);
+t_error_code	reader_from_arg(int argc, char **argv);
+t_error_code	reader(char ***ret);
+int            	is_special(char *str);
```
Comment (arikhativa) [inline include/reader.h]: int -> t_bool

File: include/reader.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# include <readline/history.h>
+
+# include "error_code.h"
+# include "libft.h"
+# include "tab.h"
+# include "macro.h"
+
+t_error_code	reader_split_arg(char *str, char ***ret);
+t_error_code	reader_from_arg(int argc, char **argv);
+t_error_code	reader(char ***ret);
```
Comment (arikhativa) [inline include/reader.h]: the name is not clear enough, add something to reader_'ACTION'()

File: include/reader.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+int            	is_special(char *str);
+t_bool      	is_dquote(char c);
+t_bool      	is_squote(char c);
+t_bool      	is_space(char c);
+
+
+t_bool			quote_is_closed(char *str);
```
Comment (arikhativa) [inline include/reader.h]: quote_is_closed -> is_quote_closed

File: src/reader/reader.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+/*   Updated: 2022/09/25 14:50:20 by alopez-g         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "reader.h"
+
+t_bool	quote_is_closed(char *str)
+{
+	char	quote_char;
+	char	i;
```
Comment (arikhativa) [inline src/reader/reader.c]: why char?
can 'i' be bigger then 127?

File: src/reader/reader.c [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	if (*(str + i) == quote_char)
+		return (i);
+	return (0);
+}
+
+t_error_code	dquote(char **line)
+{
+	char	*tmp;
+	char	*new_content;
+	char	i;
```
Comment (arikhativa) [inline src/reader/reader.c]: same 'char i'

File: src/reader/reader.c [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+				if (!(*line))
+					return (ALLOCATION_ERROR);
+				free(new_content);
+			}
+		}
+	}
+	return (SUCCESS);
+}
+
+t_error_code	reader_from_arg(int argc, char **argv)
```
Comment (arikhativa) [inline src/reader/reader.c]: who is using this func?

File: src/reader/reader.c [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+t_error_code	reader(char ***ret)
+{
+	t_error_code	err;
+	char			*line;
+
+	line = readline(MAIN_PROMPT);
+	if (!line)
+		return (ERROR);
+	dquote(&line);
```
Comment (arikhativa) [inline src/reader/reader.c]: I think this func should have a better name
but im not sure

File: src/reader/reader_split_smart.c [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+			len++;
+		*(*ret + words) = ft_substr(src, i, len);
+		if (!(*(*ret + words)))
+			return (ALLOCATION_ERROR);
+		i += is_word(src + i);
+	}
+	return (SUCCESS);
+}
+
+t_error_code	reader_split_arg(char *str, char ***ret)
```
Comment (arikhativa) [inline src/reader/reader_split_smart.c]: who is using this?

File: unit_test/reader/reader.t.c [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+/*   Created: 2022/09/25 16:25:13 by alopez-g          #+#    #+#             */
+/*   Updated: 2022/09/25 16:32:34 by alopez-g         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "unit_test_util.h"
+#include "unit_test.h"
+#include "reader.h"
+
+void	test_reader(void)
```
Comment (arikhativa) [inline unit_test/reader/reader.t.c:17]: a very cool test!

File: unit_test/reader/reader.t.c [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	CU_ASSERT(quote_is_closed("\"\""));
+	CU_ASSERT(!quote_is_closed("\""));
+
+	CU_ASSERT(cnt_words("Alvaro Lopez <Gomez>> and|Yoav|") == 9);
+	CU_ASSERT(cnt_words("|<|") == 3);
+	CU_ASSERT(cnt_words("|<<<<|") == 3);
+
+	err = tab_create(&tab, 9);
+	CU_ASSERT(err == SUCCESS);
+	err = allocate_words("Alvaro Lopez <Gomez>> and|Yoav|", &tab, 9);
```
Comment (arikhativa) [inline unit_test/reader/reader.t.c]: maybe add somthing like this:
CU_ASSERT_STRING_EQUAL(tab[0], "Alvaro")

**Review by arikhativa — CHANGES_REQUESTED** [review]
nice it looks good :)
there are a few small changes

File: include/reader.h (hunk 3 above)
Comment (al7aro) [reply include/reader.h]: reader_get_tab() ???

File: src/reader/reader.c (hunk 5 above)
Comment (al7aro) [reply src/reader/reader.c]: I changed i: char -> size_t

File: src/reader/reader.c (hunk 8 above)
Comment (al7aro) [reply src/reader/reader.c]: dquote -> close_opened_quote

**Pull request closed without merging**